File: corejs/core/promise.js
/*
 * Copyright (C) 2012 Vanderbilt University, All rights reserved.
 * 
 * Author: Miklos Maroti
 */

define([ "core/assert" ], function (ASSERT) {
	"use strict";

	var UNRESOLVED = new Error();

	var Promise = function () {
		this.value = UNRESOLVED;
		this.callback = null;
	};

	Promise.prototype.register = function (listener) {
		ASSERT(typeof callback === "function" && callback.length === 1);

		if( this.value instanceof Error ) {
			
		}
		else {
			callback(object, this.value);
		}
		if( this.value === UNRESOLVED ) {
			ASSERT(this.callback === null);
			this.callback = callback;
		}
		else if( this.value instanceof Error ) {
			throw this.value;
		}
	};

	Promise.prototype.resolve = function (value) {
		ASSERT(!(value instanceof Promise));

		if( this.callback !== null ) {
			this.callback(value);
			this.callback = null;
		}
		else {
			ASSERT(this.value === UNRESOLVED);
			this.value = value;
		}
	};

	var array = function (array) {
		ASSERT(Array.isArray(array));

		for( var i = 0; i < array.length; ++i ) {
			if( array[i] instanceof Promise ) {
				if( array[i].value !== UNRESOLVED ) {
					array[i] = array[i].value;
				}
				else {
					break;
				}
			}
		}

		if( i === array.length ) {
			return array;
		}

		var promise = new Promise(UNRESOLVED);

		var callback = function (value) {
			ASSERT(!(value instanceof Promise));

			if( value instanceof Error ) {
				promise.resolve(value);
			}
			else {
				array[i] = value;

				while( ++i < array.length ) {
					if( array[i] instanceof Promise ) {
						value = array[i].value;

						if( value instanceof Error ) {

						}
						if( value !== UNRESOLVED ) {
							array[i] = array[i].value;
						}

						array[i].register(callback);
						return;
					}
				}

				promise.resolve(array);
			}
		};

		array[i].register(callback);

		return promise;
	};

	var wrap = function (func) {
		ASSERT(typeof func === "function");

		if( func.length === 1 ) {
			return function (arg0) {
				if( arg0 instanceof Promise ) {
					if( arg0.value === UNRESOLVED ) {
						return new Call1(func, arg0);
					}
					else {
						arg0 = arg0.value;
					}
				}

				return func(arg0);
			};
		}
		else if( func.length === 2 ) {
			return function (arg0, arg1) {
				if( arg0 instanceof Promise ) {
					if( arg0.value === UNRESOLVED ) {
						return new Call2(func, 0, arg0, arg1);
					}
					else {
						arg0 = arg0.value;
					}
				}

				if( arg1 instanceof Promise ) {
					if( arg1.value === UNRESOLVED ) {
						return new Call2(func, 1, arg0, arg1);
					}
					else {
						arg1 = arg1.value;
					}
				}

				return func(arg0, arg1);
			};
		}
		else {
			// maybe throw an error
			return func;
		}
	};

	var Call1 = function (func, arg0) {
		this.value = UNRESOLVED;
		this.func = func;

		arg0.register(this);
	};

	Call1.prototype = Promise.prototype;

	var call1 = function (func, arg0) {
		ASSERT(typeof func === "function" && func.length === 1);

		if( arg0 instanceof Promise ) {
			if( arg0.value instanceof Error ) {
				return new Call1(func, arg0);
			}
			else {
				arg0 = arg0.value;
			}
		}

		return func(arg0);
	};

	return {
		array: array
	};
});

File: corejava/org/isis/ypromise/Promise.java
/*
 * Copyright (C) 2012 Vanderbilt University, All rights reserved.
 * 
 * Author: Miklos Maroti
 */

package org.isis.ypromise;

public class Promise<Type> {
	private static final Observer UNRESOLVED = new Observer() {
		public void resolved() {
			throw new IllegalStateException();
		}
	};

	protected interface Observer {
		void resolved();
	}

	private Object value;

	public Promise() {
		value = UNRESOLVED;
	}
	
	public Promise(Type value) {
		assert (!(value instanceof Exception || value instanceof Observer));
		this.value = value;
	}

	public Promise(Exception error) {
		assert(error != null);
		this.value = error;
	}
	
	public final boolean isResolved() {
		assert (value == UNRESOLVED || !(value instanceof Observer));
		return value != UNRESOLVED;
	}

	@SuppressWarnings("unchecked")
	public final Type getValue() throws Exception {
		assert (!(value instanceof Observer));

		if (value instanceof Exception)
			throw (Exception) value;
		else
			return (Type) value;
	}

	public final void setValue(Type value) {
		assert (!(value instanceof Exception || value instanceof Observer));

		setObject(value);
	}

	public final void setError(Exception error) {
		assert (error != null);
		setObject(error);
	}

	private final void setObject(Object obj) {
		Observer observer;
		synchronized (this) {
			assert (value instanceof Observer);
			observer = (Observer) value;
			value = obj;
		}

		if (observer != UNRESOLVED)
			observer.resolved();
	}

	public final void register(Observer parent) {
		assert (parent != null);

		boolean unresolved;
		synchronized (this) {
			unresolved = value == UNRESOLVED;
			if (unresolved) {
				value = parent;
			}
		}

		if (!unresolved) {
			parent.resolved();
		}
	}

	public void cancel() {
	}
}
